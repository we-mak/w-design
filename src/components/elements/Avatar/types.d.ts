import { GlobProps, SizeProps } from "../../../typings";

export interface AvatarProps extends SizeProps, GlobProps {
  // User online status
  status?: "online" | "offline" | "busy";
  avatarUrl?: string;
  // Initial content of avatar if without image
  dataInitial?: string;
  alt?: string;
}
