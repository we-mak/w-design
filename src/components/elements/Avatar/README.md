| Props       | Type   | default  | description |
| ----------- | ------ | -------- | ----------- |
| size        | string | "md"     |             |
| presence    | string | N/A      |             |
| avatarUrl   | string |          |             |
| dataInitial | string | "We mak" |             |
| alt         | string |          |             |

And `space` from **styled-system**
