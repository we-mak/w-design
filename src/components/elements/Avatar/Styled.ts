import styled from "styled-components";
import { AvatarProps } from "./types";

export const AvatarWrapper = styled.figure<AvatarProps>`
  background: #5755d9;
  border-radius: 50%;
  color: rgba(255, 255, 255, 0.85);
  display: inline-block;
  font-size: ${props => props.theme.fontSizes[1]};
  font-weight: 300;
  height: 1.6rem;
  width: 1.6rem;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
`;

export const AvatarImage = styled.img`
  border-radius: 50%;
  height: 100%;
  position: relative;
  width: 100%;
  z-index: 1;
`;
