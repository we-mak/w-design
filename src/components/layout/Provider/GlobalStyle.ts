import { createGlobalStyle } from "styled-components";
import theme from "../../../common/styleUtils/theme";
const colors = theme.colors;

/*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */
export const GlobalStyle = createGlobalStyle`
html {
  line-height: 1.15;
  -webkit-text-size-adjust: 100%;
}
body {
  margin: 0;
}
main {
  display: block;
}
h1 {
  font-size: 2em;
  margin: 0.67em 0;
}
hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible;
}
pre {
  font-family: monospace, monospace;
  font-size: 1em;
}
a {
  background-color: transparent;
}
abbr[title] {
  border-bottom: none;
  text-decoration: underline;
  text-decoration: underline dotted;
}
b,
strong {
  font-weight: bolder;
}
code,
kbd,
samp {
  font-family: monospace, monospace;
  font-size: 1em;
}
small {
  font-size: 80%;
}
sub,
sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
sub {
  bottom: -0.25em;
}
sup {
  top: -0.5em;
}
img {
  border-style: none;
}
button,
input,
optgroup,
select,
textarea {
  font-family: inherit;
  font-size: 100%;
  line-height: 1.15;
  margin: 0;
}
button,
input {
  overflow: visible;
}
button,
select {
  text-transform: none;
}
[type="button"],
[type="reset"],
[type="submit"],
button {
  -webkit-appearance: button;
}
[type="button"]::-moz-focus-inner,
[type="reset"]::-moz-focus-inner,
[type="submit"]::-moz-focus-inner,
button::-moz-focus-inner {
  border-style: none;
  padding: 0;
}
[type="button"]:-moz-focusring,
[type="reset"]:-moz-focusring,
[type="submit"]:-moz-focusring,
button:-moz-focusring {
  outline: 1px dotted ButtonText;
}
fieldset {
  padding: 0.35em 0.75em 0.625em;
}
legend {
  box-sizing: border-box;
  color: inherit;
  display: table;
  max-width: 100%;
  padding: 0;
  white-space: normal;
}
progress {
  vertical-align: baseline;
}
textarea {
  overflow: auto;
}
[type="checkbox"],
[type="radio"] {
  box-sizing: border-box;
  padding: 0;
}
[type="number"]::-webkit-inner-spin-button,
[type="number"]::-webkit-outer-spin-button {
  height: auto;
}
[type="search"] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
[type="search"]::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
details {
  display: block;
}
summary {
  display: list-item;
}
template {
  display: none;
}
[hidden] {
  display: none;
}

*,
*::before,
*::after {
  box-sizing: inherit;
}

html {
  box-sizing: border-box;
  font-size: 20px;
  line-height: 1.5;
  -webkit-tap-highlight-color: transparent;
}

body {
  font-size: .8rem;
  overflow-x: hidden;
  text-rendering: optimizeLegibility;
}


h1 {
    font-size: 1.6rem;
  }

  h2 {
    font-size: 1.4rem;
  }

  h3 {
    font-size: 1.2rem;
  }

  h4 {
    font-size: 1rem;
  }
  h5 {
    font-size: 0.9rem;
  }

  h6 {
    font-size: 0.8rem;
  }

  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    font-weight: 700;
    line-height: 1.25;
    margin-bottom: 0.5em;
    margin-top: 0;
    font-family: ${(props: any) => props.theme.fonts.medium};
  }

  h1,
  h2,
  h3,
  h4 {
    color: ${colors.N50};
  }

  h5,
  h6 {
    color: ${colors.N60};
  }

  p {
    color: ${colors.N70};
    line-height: 1.5;
    margin-bottom: 0.5em;
    margin-top: 0;
    font-family: ${(props: any) => props.theme.fonts.normal};
  }

  a {
    font-family: ${(props: any) => props.theme.fonts.normal};
    color: ${colors.B50};
    outline: none;
    text-decoration: none;

    &:hover {
      color: ${colors.B30};
    }

    &:focus,
    &:hover,
    &:active {
      text-decoration: underline;
    }
    &:focus,
    &:active {
      color: ${colors.B60};
    }
    &:focus {
      text-decoration: none;
    }
  }

  b,
  strong {
    font-weight: bolder;
  }

  small {
    font-size: 80%;
    font-weight: 400;
    color: ${colors.N30};
  }

  blockquote {
    margin-left: 0;
    padding: 0.4rem 0.8rem;
    color: ${colors.N30};
    p:last-child {
      margin-bottom: 0;
    }
  }

  table {
    border-collapse: collapse;
    border-spacing: 0;
    text-align: left;
    width: 100%;
  }
  th {
    border-bottom: 0.05rem solid ${props => props.theme.colors["N5"]};
    padding: 0.6rem 0.4rem;
    border-bottom-width: 0.1rem;
  }
  td {
    border-bottom: 0.05rem solid ${props => props.theme.colors["N2"]};
    padding: 0.6rem 0.4rem;
  }
`;
