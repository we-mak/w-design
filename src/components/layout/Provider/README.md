Theme provider is based on [styled-components](https://www.styled-components.com/) and making theme with [styled-system](http://jxnblk.com/styled-system/)

Provider provides theme for your layout, you can supply your theme by insert your theme object by `theme` props
Default theme

```js static
defaultTheme = {
  breakpoint: ["24em", "32em", "48em", "64em", "80em"],
  fonts:
    '-apple-system, BlinkMacSystemFont, Helvetica, "Helvetica Neue", system-ui, sans-serif',
  fontSizes: [
    "0.7rem",
    "0.8rem",
    "0.9rem",
    "1rem",
    "1.2rem",
    "1.4rem",
    "1.6rem",
    "1.8rem",
    "2rem"
  ],
  fontWeights = { normal: 400, strong: 500, bold: 700, xbold: 900 },
  lineHeights = [1, 1.125, 1.25, 1.5],
  letterSpacings = { normal: "normal", caps: "0.25em" },
  radii = ["0rem", "0.1rem", "0.2rem", "0.3rem", "0.4rem"],
  space = [0, 4, 8, 16, 32, 64, 128], // space is used for margin and padding scales
  shadows = [
    "none",
    `0 .2rem .5rem ${hex2Rgba(colors.N50, 0.5)}`,
    `inset 0 0 0 1px rgba(87, 85, 217, .2), 0 0 4px rgba(87, 85, 217, .2)`
  ],
  colors: {...} // see colors
};
```

```js
<Provider>
  <h1>Default </h1>
  <p>
    Lorem ipsum
    <span>
      <a href="#">dolor </a>
    </span>
    sip amet
  </p>
</Provider>
```

```js
<Provider
  theme={{
    fonts: "Roboto, Helvetica, sans-serif"
  }}
>
  <h1>Roboto</h1>
  <p>
    Lorem ipsum
    <span>
      <a href="#">dolor </a>
    </span>
    sip amet
  </p>
</Provider>
```
