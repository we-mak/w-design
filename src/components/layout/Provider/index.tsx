import * as React from "react";
import { ThemeProvider } from "styled-components";
import defaultTheme from "../../../common/theme";
import { ResetCss } from "./ResetCss";
import Root from "./Root";

export interface Props extends React.HTMLProps<HTMLDivElement> {
  theme?: any;
}

class Provider extends React.Component<Props> {
  stylesheet: any;

  static displayName = "WProvider";

  public static defaultProps: Props = {
    theme: {}
  };

  componentDidMount() {
    this.stylesheet = document.createElement("style");
    this.stylesheet.type = "text/css";
    // this.stylesheet.setAttribute("rel", "preload");
    // this.stylesheet.setAttribute("as", "style");
    this.stylesheet.innerHTML = ResetCss;

    if (document && document.head) {
      document.head.appendChild(this.stylesheet);
    }
  }

  componentWillUnmount() {
    if (this.stylesheet && document && document.head) {
      document.head.removeChild(this.stylesheet);
      delete this.stylesheet;
    }
  }

  render() {
    const { theme, ...props } = this.props;
    return (
      <ThemeProvider theme={{ ...defaultTheme, ...theme }}>
        <Root {...props as any} />
      </ThemeProvider>
    );
  }
}

export default Provider;
