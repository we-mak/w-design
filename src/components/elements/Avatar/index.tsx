import * as React from "react";
import { AvatarWrapper, AvatarDataInit, AvatarImage, PresenceMark } from "./Styled";
import { firstLetter } from "../../../common/helpers";
import { AvatarProps } from "./types";

const Avatar = ({
  size = "md",
  avatarUrl,
  dataInitial = "W Mak",
  alt,
  presence,
  ...rest
}: AvatarProps) => {
  return (
    <AvatarWrapper size={size} {...rest}>
      {dataInitial && <AvatarDataInit size={size}>{firstLetter(dataInitial)}</AvatarDataInit>}
      {avatarUrl && <AvatarImage src={avatarUrl} alt={alt} />}
      {presence && <PresenceMark />}
    </AvatarWrapper>
  );
};

export default Avatar;
