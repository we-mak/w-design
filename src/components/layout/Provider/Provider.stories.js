import React from "react";
import { storiesOf } from "@storybook/react";
import Provider from "./Provider";

storiesOf("components/Provider", module).addWithJSX("default Provider", () => (
  <Provider>
    <h1>Default </h1>
    <p>
      Lorem ipsum
      <span>
        <a href="#"> dolor </a>
      </span>{" "}
      sip amet
    </p>
  </Provider>
));
