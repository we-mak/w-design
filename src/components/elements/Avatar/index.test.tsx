import * as React from "react";
import { render, cleanup } from "@testing-library/react";
import "jest-styled-components";
import Avatar from "./index";
import Provider from "../../layout/Provider";

afterEach(cleanup);

describe("<AvatarGroup/>", () => {
  // Render test
  it("should renders correctly", () => {
    const { container } = render(
      <Provider>
        <Avatar
          size="md"
          alt="test"
          presence="online"
          avatarUrl="https://api.adorable.io/avatars/240/thunder"
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
