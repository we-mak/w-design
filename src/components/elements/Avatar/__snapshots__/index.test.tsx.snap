// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar/> should renders correctly 1`] = `
.c0 {
  border-radius: 50%;
  color: rgba(255,255,255,0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  height: 2.8rem;
  width: 2.8rem;
}

.c4 {
  border-radius: 50%;
  color: rgba(255,255,255,0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  height: 1.4rem;
  width: 1.4rem;
}

.c7 {
  border-radius: 50%;
  color: rgba(255,255,255,0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  height: 2rem;
  width: 2rem;
}

.c10 {
  border-radius: 50%;
  color: rgba(255,255,255,0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  height: 3.2rem;
  width: 3.2rem;
}

.c12 {
  border-radius: 50%;
  color: rgba(255,255,255,0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  height: 4rem;
  width: 4rem;
}

.c14 {
  border-radius: 50%;
  color: rgba(255,255,255,0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  height: 6rem;
  width: 6rem;
}

.c1 {
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 1rem;
}

.c5 {
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  font-size: .6rem;
}

.c8 {
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  font-size: .9rem;
}

.c11 {
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 1.8rem;
}

.c13 {
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 2rem;
}

.c15 {
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 2.8rem;
}

.c2 {
  border-radius: 50%;
  height: 100%;
  width: 100%;
  position: relative;
  z-index: 1;
}

.c3 {
  border-radius: 50%;
  bottom: 0.1rem;
  pointer-events: none;
  position: absolute;
  z-index: 30;
  background-color: #29A634;
  height: .7rem;
  width: .7rem;
  right: 0;
  border: 0.075rem solid #fff;
}

.c6 {
  border-radius: 50%;
  bottom: 0.1rem;
  pointer-events: none;
  position: absolute;
  z-index: 30;
  background-color: #60798E;
  height: .4rem;
  width: .4rem;
  right: 0;
  border: 0.075rem solid #fff;
}

.c9 {
  border-radius: 50%;
  bottom: 0.1rem;
  pointer-events: none;
  position: absolute;
  z-index: 30;
  background-color: #DC1F1F;
  height: .5rem;
  width: .5rem;
  right: 0;
  border: 0.075rem solid #fff;
}

<div>
  <figure
    class="c0"
  >
    <span
      class="c1"
    >
      Wm
    </span>
    <img
      alt="test"
      class="c2"
      data-src="https://api.adorable.io/avatars/240/thunder"
      src="data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP88R8AAvUB+VkkrXoAAAAASUVORK5CYII="
    />
    <span
      class="c3"
    />
  </figure>
  <figure
    class="c4"
    style="background-color: rgb(179, 245, 255);"
  >
    <span
      class="c5"
    >
      Wm
    </span>
    <span
      class="c6"
    />
  </figure>
  <figure
    class="c7"
    style="background-color: rgb(192, 182, 242);"
  >
    <span
      class="c8"
    >
      Wm
    </span>
    <span
      class="c9"
    />
  </figure>
  <figure
    class="c10"
    style="background-color: rgb(250, 155, 0);"
  >
    <span
      class="c11"
    >
      Wm
    </span>
  </figure>
  <figure
    class="c12"
    style="background-color: rgb(255, 131, 131);"
  >
    <span
      class="c13"
    >
      Wm
    </span>
  </figure>
  <figure
    class="c14"
    style="background-color: rgb(77, 97, 113);"
  >
    <span
      class="c15"
    >
      Wm
    </span>
  </figure>
</div>
`;
